### ISSUE TITLE
switch package name (formerly 'split switch_mod, examples and sandbox')

### ISSUE THREAD

**Issue by mfripp** [switch-model/switch#12]
As currently setup, the switch repository is designed to be used like this (I think):

```
switch/
    switch_mod/
    user_module.py
    inputs/
    outputs/
```

This has two problems:
1. Any files added within the `switch/` folder (e.g., `user_module.py`) need to be added to `switch/.gitignore`, and then `.gitignore` also has to ignore itself, or else collect a list of every module every user has added and ignored. Alternatively, users can leave this files untracked, and live with git's warning messages.
2. If you want `user_module.py` to be part of your own local repository, you are probably out of luck (since it falls within the area dedicated to the switch repository).

We have partially resolved these issues for switch-hawaii by using directory structures like this:

```
switch-hawaii-models/ (repo)
    project1/
        switch/ (repo)
            switch_mod/
        switch-hawaii/ (repo)
        user_module.py
        solve.py
```

In this setup, `switch-hawaii-models/.gitignore` is set to ignore the `switch` and `switch-hawaii` directories within it, so it only includes all the project-specific code (as desired). However, this makes the python path requirements a little complicated. I find `switch_mod` by including code in `project1/solve.py` that adds the adjacent switch directory to the python path before running. Then I can just execute `cd switch-hawaii-models/project1; python solve.py`.

This works fine, but I would now like to start using the standard switch_mod.solve module. To do that, I need to have both `project1/switch` and `project1` in my python path. It's easy to add `project1` to the path simply by cd'ing into that directory before starting. And if `switch_mod` resided at the top level instead of the second level, it would also be in the path at that point (along with `switch-hawaii`). i.e., I would like to have a directory structure like this:

```
switch-hawaii-models/ (repo)
    project1/
        switch_mod/ (repo)
        switch-hawaii/ (repo)
        user_module.py
        (solve.py no longer needed)
```

This would have a couple of additional advantages:
- users who don't want the examples don't have to download them, and 
- it would help smooth the path toward making switch a standard python package (in that case, switch would be installed within the site-packages directory, and would always be in the path, but should not include editable examples). 

Along with this, I would like to suggest renaming `switch_mod` to simply `switch`. I know it's a pain, but it would be a much more natural name for this module.

**Comment by mfripp** [comment]
Closing this issue, as I realized from separate discussions that it is fairly standard to have a wrapper folder for the project, and then a package folder inside that. setuptools now takes care of putting the package into the python path, resolving these issues. I would still like to rename `switch_mod` to `switch`, but that's another discussion.

**Comment by josiahjohnston** [comment]
The reason I avoided "switch" initially was a conflict with another python project: https://pypi.python.org/pypi/switch/1.1.0

Some python package naming guide I read at the time discouraged name conflicts for easy pip installation. I picked switch_mod as a fallback.

**Comment by mfripp** [comment]
Hmm, duplicate names could cause problems both for retrieving packages with pip and for installing both packages on the same system. Oh well, guess we have to live with 'switch_mod' for now. Do we need a new package name for setuptools too?

Matthias

**Comment by josiahjohnston** [comment]
We don't define the setuptools package, so why would that matter?

**Comment by mfripp** [comment]
The [setup.py script ](https://github.com/switch-model/switch/blob/master/setup.py)(which uses setuptools) specifies `name='SWITCH'`. I assume this name should correspond to the name used with pip and on pypi.python.org? But pypi.python.org already has a [`switch` package](https://pypi.python.org/pypi/switch). So should we change the package name (as reported to setuptools) to something else? 

We seem to be dealing with 2-3 different namespaces -- the name of the package as specified with `import` (e.g., `switch_mod`, `switch`, `pyomo`), the name as specified with `pip install` (e.g., `switch` and `Pyomo` already exist), and the name as specified in setup.py (currently `SWITCH` for ours). I know that the first two can be different, but it's nice if they're the same. I don't know if the latter two can be different, but again, it would be nice if they're the same. So maybe we should change our setup.py to specify `name='switch_mod'`, and then go ahead and register it on pypi.python.org?

**Comment by josiahjohnston** [comment]
I'm good with changing those three to switch_mod and registering with pypi. 
The console script will still be called switch, right? (line 50 of [setup.py](https://github.com/switch-model/switch/blob/master/setup.py))

**Comment by mfripp** [comment]
Yes, I think it makes sense to keep the console script as `switch`, unless we rename the whole thing. 

By the way, 6 of the first 7 hits for "switch power model" on google now refer to this model, so maybe the pressure for a name change is lower. We could say SWITCH is an acronym for "solar, wind, transmission, conventional generation and hydro", which is easy enough to explain. That leaves out storage and demand response, but we can probably live with that. 

So I could go either way on a complete name change. What do you think?

**Comment by josiahjohnston** [comment]
For the record, names we have considered are in [this google sheet](https://docs.google.com/spreadsheets/d/1z3c6Syupbg8GG79XI44_o3C8onnkcyWbXOz_yO6qsOQ/edit?usp=sharing).

As best I can remember, there was some general interest in changing the name, but the conversation ended up fizzling out. I'm open to changing the name, but don't have the energy to spearhead that effort right now. Keeping it as Switch (at least for now) sounds fine too.

**Comment by josiahjohnston** [comment]
Resolution:
We kept the name "Switch" for the overall project, and named the python package `switch_model`. It is available on pypi as `switch-model` and on conda forge as `switch_model`.
